bin/packages/build-worker.js

```diff
@@ -0,0 +1,153 @@
+
+/**
+ * Promisified sass.render.
+ *
+ * @type {Function}
+ */
+const renderSass = promisify( sass.render );
+
+/**
+ * Get the package name for a specified file
+ *
+ * @param  {string} file File name
+ * @return {string}      Package name
+ */
+function getPackageName( file ) {
+	return path.relative( PACKAGES_DIR, file ).split( path.sep )[ 0 ];
+}
+
+/**
+ * Get Build Path for a specified file.
+ *
+ * @param  {string} file        File to build
+ * @param  {string} buildFolder Output folder
+ * @return {string}             Build path
+ */
+function getBuildPath( file, buildFolder ) {
+	const pkgName = getPackageName( file );
+	const pkgSrcPath = path.resolve( PACKAGES_DIR, pkgName, 'src' );
+	const pkgBuildPath = path.resolve( PACKAGES_DIR, pkgName, buildFolder );
+	const relativeToSrcPath = path.relative( pkgSrcPath, file );
+	return path.resolve( pkgBuildPath, relativeToSrcPath );
+}
+
+/**
+ * Object of build tasks per file extension.
+ *
+ * @type {Object<string,Function>}
+ */
+const BUILD_TASK_BY_EXTENSION = {
+	async '.scss'( file ) {
+		const outputFile = getBuildPath( file.replace( '.scss', '.css' ), 'build-style' );
+		const outputFileRTL = getBuildPath( file.replace( '.scss', '-rtl.css' ), 'build-style' );
+
+		const [ , contents ] = await Promise.all( [
+			makeDir( path.dirname( outputFile ) ),
+			readFile( file, 'utf8' ),
+		] );
+
+		const builtSass = await renderSass( {
+			file,
+			includePaths: [ path.resolve( __dirname, '../../assets/stylesheets' ) ],
+			data: (
+				[
+					'colors',
+					'breakpoints',
+					'variables',
+					'mixins',
+					'animations',
+					'z-index',
+				].map( ( imported ) => `@import "${ imported }";` ).join( ' ' )	+
+				contents
+			),
+		} );
+
+		const result = await postcss( require( './post-css-config' ) ).process( builtSass.css, {
+			from: 'src/app.css',
+			to: 'dest/app.css',
+		} );
+
+		const resultRTL = await postcss( [ require( 'rtlcss' )() ] ).process( result.css, {
+			from: 'src/app.css',
+			to: 'dest/app.css',
+		} );
+
+		await Promise.all( [
+			writeFile( outputFile, result.css ),
+			writeFile( outputFileRTL, resultRTL.css ),
+		] );
+	},
+
+	async '.js'( file ) {
+		for ( const [ environment, buildDir ] of Object.entries( JS_ENVIRONMENTS ) ) {
+			const destPath = getBuildPath( file, buildDir );
+			const babelOptions = getBabelConfig( environment, file.replace( PACKAGES_DIR, '@wordpress' ) );
+
+			const [ , transformed ] = await Promise.all( [
+				makeDir( path.dirname( destPath ) ),
+				babel.transformFileAsync( file, babelOptions ),
+			] );
+
+			await Promise.all( [
+				writeFile( destPath + '.map', JSON.stringify( transformed.map ) ),
+				writeFile( destPath, transformed.code + '\n//# sourceMappingURL=' + path.basename( destPath ) + '.map' ),
+			] );
+		}
+	},
+};
+
+module.exports = async ( file, callback ) => {
+	const extension = path.extname( file );
+	const task = BUILD_TASK_BY_EXTENSION[ extension ];
+	if ( task ) {
+		await task( file );
```

Comment:
I experimented by putting some erroneous code into a file, and the result is an unhandled promise error (I suppose originating from `babel.transformFileAsync`). The taskbar ticked up to 100%, but the command didn't exit.

I wonder if some error handling could be added to the async function that this file exports, and the `callback` function provided with error details so that the parent process can handle exiting gracefully.

Reply:
> I wonder if some error handling could be added to the async function that this file exports, and the `callback` function provided with error details so that the parent process can handle exiting gracefully.

Interesting. Nice catch! We should definitely handle this gracefully. I'll take a look.

Reply:
This is handled better in c4f908b . It's not perfect in that I think the whole process should halt immediately when an error is encountered, but the nature of `worker-farm` makes this a bit difficult to implement. But at least it won't hang, nor will it show it as "Unhandled Promise exception".

Reply:
I had a bit of a play with it, and see what you mean.

I think the main concern would be that the script returns a non-zero exit code if an error occurs, as that would allow, for example, a CI job to end as early as possible.

One option might be to use `process.exitCode = 1;` in build.js if errors are encountered on a worker.

Reply:
worker-farm also has a `maxRetries` config option, which is by default `infinity`. I wonder if that should be capped, potentially at `0` since I wouldn't expect a build error to suddenly cure itself (I imagine this is for things like HTTP requests).

Reply:
> I had a bit of a play with it, and see what you mean.

For posterity's sake, and since I was unclear in my original comment, the issue is that other workers may still be running when the error is handled, and there is no easy way through `worker-farm` to end the process after the current round of worker handlers has completed. If the process is terminated early, subsequent worker finishes will loudly complain about the lack of a parent process to whom they expect to report.

Reply:
> worker-farm also has a `maxRetries` config option, which is by default `infinity`. I wonder if that should be capped, potentially at `0` since I wouldn't expect a build error to suddenly cure itself (I imagine this is for things like HTTP requests).

It's a good idea, though in practice I don't expect this would occur. From its documentation, the retries are relevant for process termination. Failed builds wouldn't fall under this classification.

Normally I'd say that `0` would be the more sensible default, but since it might risk that we appear to succeed with the build despite a worker having terminated, I'd rather it get stuck trying to build infinitely than to produce a bad build. At least then it's obvious that something is wrong.

More preferable would be to cap the retries, but also detect and fail the whole build if one of the processes terminates unexpected. It's not clear to me how this might be done though.

Reply:
> For posterity's sake, and since I was unclear in my original comment, the issue is that other workers may still be running when the error is handled, and there is no easy way through worker-farm to end the process only once the current round of worker handlers has completed. If the process is terminated early, subsequent worker finishes will loudly complain about the lack of a parent process to whom they expect to report.

`process.exitCode = 1` won't cause the process to be terminated early. It can be set in advance of the process exiting. I can push a commit demonstrating what I mean if that helps?

Reply:
> `process.exitCode = 1` won't cause the process to be terminated early. It can be set in advance of the process exiting. I can push a commit demonstrating what I mean if that helps?

Ah, my quoted text was meant as a clarification and not to dismiss, though I think I focused too much on this point to absorb the rest of your comment 😓 I agree with your assessment though, certainly the process should have a non-zero exit code if any build errors occur.

Reply:
> `process.exitCode = 1` won't cause the process to be terminated early. It can be set in advance of the process exiting. I can push a commit demonstrating what I mean if that helps?

Pushed (with co-authorship) in cb5156843. I confirmed that before this change, an error in the build would yield a status code `0`, and `1` after.

_Before:_

```
node ./bin/packages/build.js && echo $?
# ...
0
```

_After:_

```
node ./bin/packages/build.js && echo $?
# ...
1
```